### DISCUSSION TITLE
Implementation of database

### DISCUSSION THREAD

**Discussion by Paulmski** [Paulmski/Discord-Bot#42] [Ideas]
I'm beginning to think about implementing a database for study groups and users. Users and groups would have a many-to-many relationship meaning we would have a junction table to describe the relationship between the user and group table. My first stab at it looks like this:

CREATE DATABASE Bot;
USE Bot;

CREATE TABLE `users` (
ID INT NOT NULL,
PRIMARY KEY (ID)
) ENGINE=InnoDB;

-- Create procedure table
CREATE TABLE `study_groups` (
	ID int auto_increment PRIMARY KEY,
	`name` VARCHAR(15) NOT NULL,
    `owner` INT NOT NULL
) ENGINE=InnoDB;

CREATE TABLE `users_groups` (
	`user_ID` INT NOT NULL,
    `group_ID` INT NOT NULL
);

ALTER TABLE `users_groups`
	ADD PRIMARY KEY (user_ID, group_ID),
    ADD FOREIGN KEY (user_ID) REFERENCES users(ID),
    ADD FOREIGN KEY (group_ID) REFERENCES study_groups(ID);

My question is how should the interaction with the database be implemented? Should we model the database and create two classes for user and group? Should there be a separate class to make interacting with the database easier? I was thinking we'd use the library `mysql-connector-python`, but I'm also open to suggestions.

**Comment by tuanyvan** [comment]
MySQL databases would be a solid implementation for functionalities like study groups. Keeping track of newly made private groups, their IDs, and their users in variables would only work if the bot never turned off. Databases allow the bot to have a form of long-term memory.

I think the best approach is storing the database & tables on a local `.db` file, meaning we don't need to rent processing space externally; it can be done on the `localhost`.

By the nature of `.db` files, I believe we can just create the database with blank tables. When the bot runs, it will append and delete rows from the three tables (users, study_groups, and users_groups), and if the bot needs to turn off, the data will still be there. The `mysql-connector-python` library should give us everything we need to insert and delete values.

We should definitely create a class with methods for interacting with the database to abstract it away from `__main__.py`. It can take the required user/password/host/database credentials as its instance variables. Its behavior should be able to create groups, delete groups, add members to groups, remove members from groups, and delete groups after inactivity. It should make use of DML commands to the `.db` file for each of these interactions too.

I'm not too sure if a user/group class needs to be created since user ID and private group ID can all be accessed from the `ctx` Context object provided by commands.

**Reply by Paulmski** [reply]
Definitely agree, hosting it on the local host would be the best option. 

After thinking more about implementing study groups I've realized that it can be done without a database. The study groups can be made on-demand in a category called "study-groups", each group is made private so only people mentioned in the create command can join. There will also need to be other commands to delete study groups, and add other members (after the group is created).


The bot can then iterate through each study-groups channel look at the last sent message time stamp and if it is greater than x days the voice and text channel is removed.

**Reply by tuanyvan** [reply]
This approach sounds simpler than creating a database + connection.

Having group members join only by the creator's invitation also prevents random users from simply 'joining-spying-and-leaving', something the initial "emote-to-join" system did not account for.
